File: Project/NetworkingManaged/TCPServerSocket.cs
﻿// Copyright 2019. All Rights Reserved.
using GameFramework.BinarySerializer;
using GameFramework.Common.Timing;
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;

namespace GameFramework.Networking
{
	public class TCPServerSocket : ServerSocket
	{
		private class TCPClient : Client
		{
			private IPEndPoint endPoint;

			public override bool IsReady
			{
				get { return (SocketUtilities.GetIsReady(Socket) && base.IsReady); }
			}

			public Socket Socket
			{
				get;
				private set;
			}

			public override IPEndPoint EndPoint
			{
				get { return endPoint; }
			}

			public TCPClient(Socket Socket)
			{
				this.Socket = Socket;
				endPoint = (IPEndPoint)Socket.RemoteEndPoint;
			}
		}

		private class TCPClientList : List<TCPClient>
		{ }

		private TCPClientList clients = null;

		public uint MaxConnection
		{
			get;
			set;
		}

		public override Client[] Clients
		{
			get
			{
				lock (clients)
					return clients.ToArray();
			}
		}

		public TCPServerSocket(uint MaxConnection = 32) : base(Protocols.TCP)
		{
			clients = new TCPClientList();

			this.MaxConnection = MaxConnection;
		}

		public override void UnBind()
		{
			for (int i = 0; i < clients.Count; ++i)
				SocketUtilities.CloseSocket(clients[i].Socket);

			clients.Clear();

			base.UnBind();
		}

		public override void DisconnectClient(Client Client)
		{
			lock (clients)
				if (clients.Contains((TCPClient)Client))
					clients.Remove((TCPClient)Client);

			base.DisconnectClient(Client);
		}

		public override void Listen()
		{
			Socket.Listen((int)MaxConnection);

			base.Listen();
		}

		public virtual void Send(Client Target, byte[] Buffer)
		{
			Send(Target, Buffer, 0, (uint)Buffer.Length);
		}

		public virtual void Send(Client Target, byte[] Buffer, uint Length)
		{
			Send(Target, Buffer, 0, Length);
		}

		public virtual void Send(Client Target, byte[] Buffer, uint Index, uint Length)
		{
			BufferStream buffer = Packet.CreateOutgoingBufferStream(Length);

			buffer.WriteBytes(Buffer, Index, Length);

			SendInternal(Target, buffer);
		}

		protected virtual void SendInternal(Client Client, BufferStream Buffer)
		{
			AddSendCommand(new ServerSendCommand(Client, Buffer, Timestamp));
		}

		protected override void AcceptClients()
		{
			try
			{
				Socket clientSocket = Socket.Accept();

				TCPClient client = new TCPClient(clientSocket);
				client.Statistics.SetPacketCountRate(PacketCountRate);

				lock (clients)
					clients.Add(client);

				RaiseOnClientConnected(client);
			}
			catch (SocketException e)
			{
				if (e.SocketErrorCode != SocketError.WouldBlock)
					throw e;
			}
		}

		protected override void ReadFromClients()
		{
			lock (clients)
			{
				TCPClientList disconnectedClients = new TCPClientList();

				for (int i = 0; i < clients.Count; ++i)
				{
					TCPClient client = clients[i];

					try
					{
						int receiveSize = 0;
						int availableSize = client.Socket.Available;

						lock (client.Socket)
						{
							if (availableSize == 0)
							{
								if (!client.IsReady)
								{
									disconnectedClients.Add(client);

									HandleClientDisconnection(client);
								}

								continue;
							}

							receiveSize = client.Socket.Receive(ReceiveBuffer, (int)ReceiveBufferIndex, availableSize, SocketFlags.None);
						}

						client.Statistics.AddReceivedPacketFromLastSecond();

						ProcessReceivedBuffer(client, (uint)receiveSize);
					}
					catch (SocketException e)
					{
						if (e.SocketErrorCode == SocketError.WouldBlock)
							continue;
						else if (e.SocketErrorCode == SocketError.ConnectionReset)
						{
							disconnectedClients.Add(client);

							HandleClientDisconnection(client);

							continue;
						}

						throw e;
					}
					catch (Exception e)
					{
						throw e;
					}
				}

				for (int i = 0; i < disconnectedClients.Count; ++i)
					clients.Remove(disconnectedClients[i]);
			}
		}

		protected override void HandleIncomingBuffer(Client Client, BufferStream Buffer)
		{
			byte control = Buffer.ReadByte();

			double time = Time.CurrentEpochTime;

			Client.Statistics.SetLastTouchTime(time);

			if (control == Constants.Control.BUFFER)
			{
				BufferStream buffer = Packet.CreateIncomingBufferStream(Buffer.Buffer);

				ProcessReceivedBuffer(Client, buffer);
			}
			else if (control == Constants.Control.PING)
			{
				double sendTime = Buffer.ReadFloat64();

				Client.Statistics.SetLatency((uint)((time - sendTime) * 1000));

				BufferStream pingBuffer = Packet.CreatePingBufferStream();

				SendInternal(Client, pingBuffer);
			}
		}

		protected override bool HandleSendCommand(SendCommand Command)
		{
			ServerSendCommand sendCommand = (ServerSendCommand)Command;
			TCPClient client = (TCPClient)sendCommand.Client;

			//if (!client.IsReady)
			//	return false;

			client.Statistics.AddBandwidthOut(Command.Buffer.Size);

			client.Statistics.SetLastTouchTime(Time.CurrentEpochTime);

			if (!SendOverSocket(client.Socket, Command.Buffer))
			{
				HandleClientDisconnection(client);

				lock (clients)
				{
					clients.Remove(client);
				}

				return false;
			}

			return true;
		}

		protected override void ProcessReceivedBuffer(Client Sender, BufferStream Buffer)
		{
			HandleReceivedBuffer(Sender, Buffer);
		}

		protected override void CloseClientConnection(Client Client)
		{
			base.CloseClientConnection(Client);

			SocketUtilities.CloseSocket(((TCPClient)Client).Socket);
		}

		public static void OpenDynamicTCPPorts(ushort From = 1500, ushort Count = 63000)
		{
			SocketUtilities.OpenDynamicTCPPorts(From, Count);
		}
	}
}

File: Project/NetworkingNative/Compile/TCPServerSocket.cpp
// Copyright 2019. All Rights Reserved.
#include "..\Include\TCPServerSocket.h"
#include "..\Include\Packet.h"
#include "..\Include\Constants.h"
#include <Timing\Time.h>

using namespace GameFramework::Common::Timing;

namespace GameFramework::Networking
{
	TCPServerSocket::TCPServerSocket(uint32_t MaxConnection) :
		ServerSocket(PlatformNetwork::IPProtocols::TCP),
		m_MaxConnection(MaxConnection)
	{
	}

	void TCPServerSocket::UnBind(void)
	{
		m_Clients.clear();

		ServerSocket::UnBind();
	}

	void TCPServerSocket::DisconnectClient(Client* const Client)
	{
		WAIT_FOR_BOOL(m_ClientsLock);

		m_Clients.remove(reinterpret_cast<TCPClient*>(Client));

		ServerSocket::DisconnectClient(Client);
	}

	void TCPServerSocket::Listen(void)
	{
		SocketUtilities::Listen(GetSocket(), m_MaxConnection);

		ServerSocket::Listen();
	}

	void TCPServerSocket::Send(const Client* Target, byte* const Buffer, uint32_t Length)
	{
		Send(Target, Buffer, 0, Length);
	}

	void TCPServerSocket::Send(const Client* Target, byte* const Buffer, uint32_t Index, uint32_t Length)
	{
		BufferStream buffer = Packet::CreateOutgoingBufferStream(Length);

		buffer.WriteBytes(Buffer, Index, Length);

		AddSendCommand(const_cast<Client*>(Target), buffer);
	}

	void TCPServerSocket::AddSendCommand(Client* Target, const BufferStream& Buffer)
	{
		BaseSocket::AddSendCommand(new ServerSendCommand(Target, Buffer, GetTimestamp()));
	}

	void TCPServerSocket::AcceptClients(void)
	{
		try
		{
			Socket clientSocket = 0;
			IPEndPoint endPoint(IPAddress::AnyV6, 0);
			if (SocketUtilities::Accept(GetSocket(), clientSocket, endPoint))
			{
				TCPClient* client = new TCPClient(clientSocket, endPoint);
				client->GetStatistics().SetPacketCountRate(GetPacketCountRate());

				{
					WAIT_FOR_BOOL(m_ClientsLock);

					m_Clients.push_back(client);
				}

				RaiseOnClientConnected(client);
			}
		}
		catch (PlatformNetwork::SocketException e)
		{
			if (e.GetError() != PlatformNetwork::Errors::WouldBlock)
				throw e;
		}
		catch (exception e)
		{
			throw e;
		}
	}

	void TCPServerSocket::ReadFromClients(void)
	{
		WAIT_FOR_BOOL(m_ClientsLock);
		TCPClientList disconnectedClients;

		std::byte* receiveBuffer = GetReceiveBuffer();

		for (auto client : m_Clients)
		{
			try
			{
				Socket clientSocket = client->GetSocket();

				uint32_t availableSize = SocketUtilities::GetAvailableBytes(clientSocket);
				if (availableSize == 0)
				{
					if (!client->GetIsReady())
						disconnectedClients.push_back(client);

					continue;
				}

				uint32_t receiveSize = availableSize;
				if (!SocketUtilities::Receive(clientSocket, receiveBuffer, GetReceiveBufferIndex(), receiveSize))
					continue;

				ServerSocket::ProcessReceivedBuffer(client, receiveSize);
			}
			catch (PlatformNetwork::SocketException e)
			{
				if (e.GetError() == PlatformNetwork::Errors::WouldBlock)
					continue;
				if (e.GetError() == PlatformNetwork::Errors::ConnectionReset)
				{
					disconnectedClients.push_back(client);

					continue;
				}

				throw e;
			}
			catch (exception e)
			{
				throw e;
			}
		}

		for (auto client : disconnectedClients)
		{
			m_Clients.remove(client);

			HandleClientDisconnection(client);
		}
	}

	void TCPServerSocket::HandleIncomingBuffer(Client* Client, BufferStream& Buffer)
	{
		byte control = Buffer.ReadByte();

		double time = Time::GetCurrentEpochTime();

		Client->GetStatistics().SetLastTouchTime(time);

		if (control == Constants::Control::BUFFER)
		{
			BufferStream buffer = Packet::CreateIncomingBufferStream(Buffer.GetBuffer(), Buffer.GetSize());

			ProcessReceivedBuffer(Client, buffer);
		}
		else if (control == Constants::Control::PING)
		{
			double sendTime = Buffer.ReadFloat64();

			Client->GetStatistics().SetLatency(abs(time - sendTime) * 1000);

			BufferStream pingBuffer = Packet::CreatePingBufferStream();

			AddSendCommand(Client, pingBuffer);
		}
	}

	bool TCPServerSocket::HandleSendCommand(SendCommand* Command)
	{
		ServerSendCommand* sendCommand = reinterpret_cast<ServerSendCommand*>(Command);
		TCPClient* client = reinterpret_cast<TCPClient*>(sendCommand->GetClient());

		//if (!client->GetIsReady())
		//	return false;

		client->GetStatistics().AddBandwidthOut(Command->GetBuffer().GetSize());

		client->GetStatistics().SetLastTouchTime(Time::GetCurrentEpochTime());

		if (!BaseSocket::SendOverSocket(client->GetSocket(), Command->GetBuffer()))
		{
			HandleClientDisconnection(client);

			WAIT_FOR_BOOL(m_ClientsLock);
			m_Clients.remove(client);

			return false;
		}

		return true;
	}

	void TCPServerSocket::ProcessReceivedBuffer(Client* Sender, BufferStream& Buffer)
	{
		HandleReceivedBuffer(Sender, Buffer);
	}

	void TCPServerSocket::CloseClientConnection(Client* Client)
	{
		ServerSocket::CloseClientConnection(Client);

		SocketUtilities::CloseSocket(reinterpret_cast<TCPClient*>(Client)->GetSocket());

		delete Client;
	}
}